datanode/src/storage_api.py

```diff
@@ -292,17 +297,86 @@ def GridCellsOperationDataHandler(zoom, gufi):
   return _FormatResult(result)
 
 
+@webapp.route(
+  '/UVR/<zoom>/<message_id>',
+  methods=['PUT', 'POST', 'DELETE'])
+def UvrDataHandler(zoom, message_id):
+  """Handles the web service request for emplacing or removing a UVR.
+
+  Args:
+    zoom: zoom level in slippy tile format
+    message_id: unique message ID of constraint
+    OAuth access_token as part of the header
+  Returns:
+    200 with token and metadata in JSON format,
+    or the nominal 4xx error codes as necessary.
+  """
+  uss_id = _ValidateAccessToken(UVR_WRITE_SCOPE)
+
+  if not uss_id:
+    return {
+      'status': 'fail',
+      'code': status.HTTP_400_BAD_REQUEST,
+      'message': 'uss_id must be provided in the request to modify a UVR.'}
+
+  uvr_json = _GetRequestParameter('uvr', None)
+  if not uvr_json:
+    return {
+      'status': 'error',
+      'code': status.HTTP_400_BAD_REQUEST,
+      'message': 'uvr field must be provided in the form data request'}
+
+  result = {}
+  try:
+    unvalidated_uvr = json.loads(uvr_json)
+  except ValueError as e:
+    abort(status.HTTP_400_BAD_REQUEST, 'Error parsing UVR JSON: ' + e.message)
+
+  try:
+    uvr = uvrs.Uvr(unvalidated_uvr)
+
+    if uvr['message_id'] != message_id:
+      raise ValueError('message_id "%s" in request does not match message_id '
+                       '"%s" in UVR' % (message_id, uvr['message_id']))
+    if uvr['originator_id'] != uss_id:
+      raise ValueError('authenticated client ID "%s" does not match '
+                       'originator_id "%s" in UVR' % (uss_id,
+                                                      uvr['originator_id']))
+
+    zoom = int(zoom)
+    tiles = uvr.get_tiles(zoom)
+    if len(tiles) > UVR_TILE_LIMIT:
+      raise OverflowError('Limit of %d tiles impacted exceeded (%d)'
+                          % (UVR_TILE_LIMIT, len(tiles)))
+  except (KeyError, ValueError, TypeError) as e:
+    abort(status.HTTP_400_BAD_REQUEST, e.message)
+  except OverflowError as e:
+    abort(status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, e.message)
+  if request.method in ('PUT', 'POST'):
+    result = wrapper.insert_uvr(zoom, tiles, uvr)
+  elif request.method == 'DELETE':
+    result = wrapper.delete_uvr(zoom, tiles, uvr)
```

Comment:
Shouldn't delete only require a UVR identifier? This makes the API straight forward with nothing in the body, and we want to make removal as easy as possible (consider failover, expiration, etc.)

Reply:
No, at very minimum it needs the geographic extents of the UVR to find the appropriate grid cells.  We could have the coordinates as a PUT field like GridCellsOperator, but 1) that requires the client to convert between coordinates formats and 2) makes the likelihood of incomplete delete much higher.

If an emplacing client has lost the full UVR definition but still has the ```message_id``` and at least one grid cell containing it, they can easily retrieve the UVR and then parrot it back for deletion.  If they don't know at least one grid cell, then the UVR is essentially orphaned -- it's the same problem as with orphaned Operators.  If we require just the coordinates (in a different format) in the DELETE request, a client who has lost the full UVR definition will still need to query the grid cell to look up the coordinates, and additionally they will have to perform the conversion.

Reply:
Roger, makes sense.